### PR TITLE
Webb Mixer Extended (follow-up grant)

### PR DESCRIPTION
**Pull request by drewstone**
# Grant Application Checklist

- [x] The [application template](https://github.com/w3f/Open-Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [x] A BTC or Ethereum (DAI) address for the payment of the milestones is provided inside the application.  
- [x] I have read and acknowledged the [Terms and Conditions](https://github.com/w3f/Open-Grants-Program/blob/master/src/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The total funding amount of the project is below USD $30k for initial grant applications and $100k for follow-up grants.
- [x] The initial PR contains only one commit (squash if needed before submitting your PR).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery).

### CODE REVIEW

**Comment by alxs** [comment]
Thanks for the application @drewstone. The total funding amount does need to be below $100k for all grants, would this be possible with the proposed scope or would you rather remove a milestone/some of the deliverables? You can always apply for a further grant after completion.

**Comment by drewstone** [comment]
Ok @alxs 👍

**Comment by mmagician** [comment]
@drewstone Since you haven't adjusted the application, it wasn't entirely clear whether you're ok with lowering the total cost or removing some of the deliverables?

**Comment by drewstone** [comment]
Oops, it seems I didn't push the changes I thought I had.

**Review by mmagician — APPROVED** [review]

**Comment by Noc2** [comment]
@drewstone are you aware of https://github.com/patractlabs/zkmega, I guess you can potentially reuse some of their work?

**Comment by drewstone** [comment]
@Noc2 thanks for sharing, though it looks like a completely different implementation of tools we also have built/are working with. Perhaps in the future we can work on a grant to extend our interfaces to support theirs as well but I feel arkworks is a much better standard to be using since it spans outside the Polkadot ecosystem.

I don't think we will be using any of that for this grant however.

**Comment by drewstone** [comment]
@Noc2 I see some of arkworks tests in their repo. Thanks for sharing for sure! I will keep an eye on the project. For us however, we're building quite circuit-specific verifiers atm (for mixers/merkle tree proofs) and so not sure immediately how we'd jump into that project or whether they could similarly integrate with ours: https://github.com/webb-tools/anon/

**Review by Noc2 — APPROVED** [review]
Sorry for the late reply here. Makes sense. I’m happy to support the project. Btw. anonima also sounds like an interesting project and it seems some of the links on your website don’t work yet. I would be happy to follow your team on twitter ;-)

**Comment by drewstone** [comment]
thanks @Noc2 ! Here's our twitter: https://twitter.com/webbprotocol

**Comment by drewstone** [comment]
Would there be someone at W3F interested in talking about `anonima`? There's a lot of deeper privacy tech we'd be interested in building out for Substrate potentially, such as PIR (private information retrieval) if there were larger/longer grants available @Noc2.

File: applications/MIXERv2.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+| 3. | Docker | We will provide a dockerfile to demonstrate the full functionality of our chain |
+
+# Additional task details
+
+**Pallet:**
+
+- [ ] Create generalised structure for integrating new zero-knowledge backends into the mixer.
+    - [ ]  Add support to the merkle tree pallet for circuit specific setups using arbitrary backends (Groth16, Bulletproofs)
+    - [ ]  Augment mixer data structure with setup/backend.
+    - [ ]  Tests
```
Comment (burdges) [inline applications/MIXERv2.md:81]: zk storage is important.  I'm quite happy with this goal, even ignoring everything else.

File: applications/MIXERv2.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    - [ ] Add pallet-asset style support
+    - [ ] Augment mixer creation flow to support creation for arbitrary token types and denominations.
+
+**Circuit**
+
+- [ ] A hash preimage circuit
+- [ ] A non-zero circuit and bounded number circuit
+- [ ] A merkle tree proof verification circuit
+- [ ] Support for fees inside these circuits.
+- [ ] All using `arkworks`
```
Comment (burdges) [inline applications/MIXERv2.md:93]: We'll want zk friendly hashing here, like Poseidon, but I think Drew knows that.

Comment (mmagician) [reply applications/MIXERv2.md:93]: Yes, they had already used Poseidon for the first grant

**Review by alxs — APPROVED** [review]

**Comment by drewstone** [comment]
@alxs @Noc2 we've finished the first milestone and would be happy to start the delivery/review process. Here is the blog post to accompany it: https://www.notion.so/hicommonwealth/Blog-post-268628ee32ac4f21b9b8531385b00458.

Let me know what the steps are again.

**Comment by drewstone** [comment]
Also we learned a lot building out this project that may require us to work more intensely on the pallets rather than jumping into the UI work. Is it ok if I re-evaluate the 2nd milestone and consider proposing a new grant for improving the pallets? Namely, parameter storage requires a lot of work on its own and I think we can improve it to be much more efficient.

**Comment by alxs** [comment]
@drewstone good to hear that, thanks for the quick delivery! Please follow the instructions on the [milestone delivery repo](https://github.com/w3f/Grant-Milestone-Delivery#mailbox-milestone-delivery-process) in order to submit it. Also feel free to send the article to grantsPR@web3.foundation for proofreading and cross-promotion. You're welcome to submit an amendment for the second milestone (just modify the existing file via PR). The total price should still stay below $100k and the application will need to be reevaluated, but you can certainly do that and then apply for another follow-up grant once you've completed this one. Independently of this, could you send us an email to grants@web3.foundation to have a chat about your work and mid- to longer-term plans (maybe once you've submitted)?

**Comment by drewstone** [comment]
Hey @alxs sorry about this! It is actually done. We just need to wrap up submission.

**Comment by drewstone** [comment]
Hey @alxs please refer to our dApp for instructions on what to run to see everything in action

https://github.com/webb-tools/webb-dapp

It includes documentation on how to run everything relevant for a Substrate mixer + dApp + relayer.
- The dApp
- Substrate mixer
- Transaction relayer
- API/SDK integrated throughout

**Pull request merged**